controllers/vspherevm_controller.go

```diff
@@ -341,7 +346,14 @@ func (r vmReconciler) reconcileNormal(ctx *context.VMContext) (reconcile.Result,
 	}
 
 	// Update the VSphereVM's network status.
-	r.reconcileNetwork(ctx, vm)
+	if ok, err := r.reconcileNetwork(ctx, vm); !ok {
+		if err != nil {
+			return reconcile.Result{}, errors.Wrapf(err,
+				"unexpected error while reconciling network for %s", ctx)
+		}
+		ctx.Logger.Info("network is not reconciled")
+		return reconcile.Result{RequeueAfter: 30 * time.Second}, nil
```

Comment:
This does change the behavior of when the VSphereVM's status.ready field is set to true, along with the provisioned condition. It now waits until guest tools reports at least one IP address before doing this. Will this have any negative repercussions with the delay in setting ready/provisioned? Before, CAPV considered the VSphereVM and therefore the VSphereMachine as ready without needing guest tools to report an IP. Now, this adds delays in 30-second increments.

Do you have a strong need to delay considering the VSphereVM/VSphereMachine ready until guest tools reports an IP?

I'm tempted to say we don't include this change as part of this PR, and instead only do the `isWaitingForStaticIPAllocation` part. WDYT?

Reply:
No, there is no specific requirement to delay setting the VSphereVM to 'Ready'. Without the 'Requeue', for some reason the 'PollImmediateInfinite' is indefinitely stuck at "resource is not patched". I need to figure out the root cause. Otherwise, we can restrict this PR to only do the "isWaitingForStaticIPAllocation".

Reply:
just my understanding, let me know if it's correct. without this requeue, since vspheremachine controller is waiting for IP in vspherevm.status, so until next resync reconciliation loop (default 10m), vspheremachine will not be ready. 

for the same reason, even line #320, we may need to do requeue as well.  @abhinavnagaraj did some tests, it took 20m for LB+CP to come up. 

Unless there is other places that we need to change to wait for the IPs synchronously.

Reply:
I'm about to run into a meeting, but I think the vmService.ReconcileVM() call which is above this line kicks off a goroutine that waits for all IPs to be seen in the VM, and then it reenqueues the VSphereVM. I'm surprised this isn't working and is causing the resync delay... or maybe I'm misreading it. I'll check more later when I have more time. Thanks.

Reply:
Thanks. I'll double check.

Reply:
We figured the issue is with line https://github.com/kubernetes-sigs/cluster-api-provider-vsphere/blob/4ee7445b4b9fcebc660a486a40bb05105a239c46/pkg/services/govmomi/util.go#L443. Since static IP is required to be in the CIDR format(https://github.com/kubernetes-sigs/cluster-api-provider-vsphere/issues/854), we need to parse the 'specIP' which is in CIDR format here, before comparing it with the discoveredIP. I was able to verify the PR after fixing this.

Reply:
Great find!